File: ui/drawings/fib_math.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class FibLevel:
    ratio: float
    price: float


def build_retracement_levels(start_price: float, end_price: float, levels: list[float]) -> list[FibLevel]:
    return [
        FibLevel(ratio=float(level), price=float(end_price + (start_price - end_price) * level))
        for level in levels
    ]


def build_extension_levels(a_price: float, b_price: float, c_price: float, levels: list[float]) -> list[FibLevel]:
    delta = b_price - a_price
    return [
        FibLevel(ratio=float(level), price=float(c_price + delta * level))
        for level in levels
    ]

File: native/drawing/src/FibMath.cpp
#include "tradereview/drawing/FibMath.h"

namespace tradereview::drawing {

std::vector<FibLevel> build_retracement_levels(
    double start_price,
    double end_price,
    const std::vector<double>& levels)
{
    std::vector<FibLevel> rows;
    rows.reserve(levels.size());
    for (const auto level : levels) {
        rows.push_back(FibLevel{
            level,
            end_price + ((start_price - end_price) * level),
        });
    }
    return rows;
}

std::vector<FibLevel> build_extension_levels(
    double a_price,
    double b_price,
    double c_price,
    const std::vector<double>& levels)
{
    const auto delta = b_price - a_price;
    std::vector<FibLevel> rows;
    rows.reserve(levels.size());
    for (const auto level : levels) {
        rows.push_back(FibLevel{
            level,
            c_price + (delta * level),
        });
    }
    return rows;
}

} // namespace tradereview::drawing
